src/transformers/trainer.py

```diff
@@ -666,3 +723,19 @@ def _prediction_loop(
             metrics["loss"] = np.mean(eval_losses)
 
         return PredictionOutput(predictions=preds, label_ids=label_ids, metrics=metrics)
+
+    def distributed_concat(self, arr: np.ndarray, num_total_examples: int) -> np.ndarray:
+        assert self.args.local_rank != -1
+
+        def create_copy():
+            return torch.from_numpy(arr).to(self.args.device)
+
+        output_tensors = [create_copy() for _ in range(torch.distributed.get_world_size())]
+        torch.distributed.all_gather(output_tensors, create_copy())
+
+        concat = torch.cat(output_tensors, dim=0)
+        output = concat.detach().cpu().numpy()
+
+        # truncate the dummy elements added by SequentialDistributedSampler
+        output = output[:num_total_examples]
+        return output
```

Comment:
My suggestion would be to not convert it to a numpy array beforehand (so save predictions as tensors but still detach and move to cpu). Then this function can become something like:

```python
def distributed_concat(self, x: torch.Tensor, num_total_examples: int) -> np.ndarray:
    gather = [torch.empty_like(x) for _ in range(dist.get_world_size())]
    dist.all_gather(gather, x)
    return torch.cat(gather)[:num_total_examples].numpy()
````

Or an alternative that uses `gather` rather than `gather_all` so that we only gather all results to the master process.

Reply:
I agree on the first point but it involves more rewriting that's affecting all code paths so I was a bit reluctant to do it. Might take a stab though...

For gather, I don't think nccl supports it. https://pytorch.org/docs/stable/distributed.html

Reply:
Sure, whatever you think fits best in the whole thing. Ultimately, I suppose you'd want to keep the tensors (detached) on CPU as long as you can so that you are free in using them however you want. 

You are absolutely correct about NCCL and `gather()`!